### ISSUE TITLE
useWatch and reset produces too many renders (used in form array)

### ISSUE THREAD

**Issue by eugenegordeiev** [react-hook-form/react-hook-form#3683]
**Describe the bug**
When building a form which consists of fields array in combination of useWatch() and reset(), the reset introduces 3 additional renders for every array item.  The more fields added to the array, the more re-renders it does to a point where the page is extremely slow.

**To Reproduce**

https://codesandbox.io/s/rhf-watch-examples-v6rc4-forked-lzikg?file=/src/App.tsx

In this example I use 2 components.  App component which initializes the form and adds fields in a loop by including an ProfileEditContactsItem component.

ProfileEditContactsItem is using useWatch() to watch a field in the form (this is needed for additional functionality which is not included in this example).

When submit button is pressed (in App component), it triggers a reset() and the reset() re-render the APP component twice and 6 times the ProfileEditContactsItem component (see console messages).

With every additional item added to the array, it adds additional 1 re-render to APP and 2 for ProfileEditContactsItem component.

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

- OS: OSX
- Browser: chrome
- Version: 87

I dont think this is whats expected from useWatch() and reset().

**Comment by bluebill1049** [comment]
That's the expected re-render amount. 

- useWatch only re-render what's inside the component
- reset will re-render the entire component

**Comment by bluebill1049** [comment]
> The more fields added to the array, the more re-renders it does to a point where the page is extremely slow.

can you share a codesandbox on the above?

**Comment by bluebill1049** [comment]
<img width="807" alt="Screen Shot 2020-12-15 at 8 46 28 am" src="https://user-images.githubusercontent.com/10513364/102139503-07bf2400-3eb2-11eb-8125-df0ba6be6cc7.png">

why are you invoking `getValues` on each render?

**Comment by eugenegordeiev** [comment]
> That's the expected re-render amount.
> 
> * useWatch only re-render what's inside the component
> * reset will re-render the entire component

That is was what I assumed.. so why with every additional item in the array and APP is re-rendered again?  I would assume the App (main) component should re-render once and useWatch would re-render every sub-component (once).

You can try adding more items to the defaultValues contacts array to see the difference.

**Comment by eugenegordeiev** [comment]
> <img alt="Screen Shot 2020-12-15 at 8 46 28 am" width="807" src="https://user-images.githubusercontent.com/10513364/102139503-07bf2400-3eb2-11eb-8125-df0ba6be6cc7.png">
> 
> why are you invoking `getValues` on each render?

getValues() does not seem to impact the number of re-renders.

**Comment by bluebill1049** [comment]
> > <img alt="Screen Shot 2020-12-15 at 8 46 28 am" width="807" src="https://user-images.githubusercontent.com/10513364/102139503-07bf2400-3eb2-11eb-8125-df0ba6be6cc7.png">
> > why are you invoking `getValues` on each render?
> 
> getValues() does not seem to impact the number of re-renders.

You talking about perf here right? getValues() is not for free.

**Comment by bluebill1049** [comment]
> > That's the expected re-render amount.
> > 
> > * useWatch only re-render what's inside the component
> > * reset will re-render the entire component
> 
> That is was what I assumed.. so why with every additional item in the array and APP is re-rendered again? I would assume the App (main) component should re-render once and useWatch would re-render every sub-component (once).
> 
> You can try adding more items to the defaultValues contacts array to see the difference.

How would it update field array if not re-render the app?

**Comment by eugenegordeiev** [comment]
Sorry, I actually found the problem.  I've added the following:
      <input
        type="hidden"
        ref={register}
        value={type?.icon ? type.icon : ""}
        name={`contacts[${index}].icon`}
      />

The input adds more re-renders in my local, however in codesandbox it does not have any effect.  So I've changed the <input> to a <Controller> and now the number of re-renders always stay the same regardless of how many elements are there in the array.  In any case it looks like the issue is on my local only.

Thanks for you help.
